File: services/audit-service/src/audit/audit.service.ts
import { randomUUID } from 'node:crypto';

import { Injectable, Logger, NotFoundException } from '@nestjs/common';

import { DatabaseService } from '../database/database.service';
import { RequestContextService } from '../observability/request-context.service';
import {
  ComplianceReportQueryDto,
  ExportAuditDto,
  QueryAuditEventsDto,
  RecordAuditEventDto,
} from './dto/audit.dto';

export interface AuditRow {
  id: string;
  actor_id: string;
  action: string;
  resource: string;
  resource_id: string | null;
  tenant_id: string | null;
  metadata: Record<string, unknown>;
  created_at: Date;
}

@Injectable()
export class AuditService {
  private readonly logger = new Logger(AuditService.name);

  constructor(
    private readonly db: DatabaseService,
    private readonly requestContext: RequestContextService,
  ) {}

  async recordEvent(dto: RecordAuditEventDto): Promise<AuditRow> {
    const tenantId = dto.tenantId ?? this.requestContext.tenantId ?? 'default';
    const id = randomUUID();

    const result = await this.db.query<AuditRow>(
      `INSERT INTO audit_events (id, actor_id, action, resource, resource_id, tenant_id, metadata, created_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7::jsonb, NOW()) RETURNING *`,
      [
        id,
        dto.actorId,
        dto.action,
        dto.resource,
        dto.resourceId ?? null,
        tenantId,
        JSON.stringify(dto.metadata ?? {}),
      ],
    );

    this.logger.log({ id, action: dto.action, resource: dto.resource }, 'Audit event recorded');
    return result.rows[0]!;
  }

  async findAll(query: QueryAuditEventsDto) {
    const page = query.page ?? 1;
    const pageSize = query.pageSize ?? 20;
    const offset = (page - 1) * pageSize;

    const conditions: string[] = [];
    const params: unknown[] = [];
    let idx = 1;

    if (query.actorId) {
      conditions.push(`actor_id = $${idx++}`);
      params.push(query.actorId);
    }
    if (query.action) {
      conditions.push(`action = $${idx++}`);
      params.push(query.action);
    }
    if (query.resource) {
      conditions.push(`resource = $${idx++}`);
      params.push(query.resource);
    }
    if (query.resourceId) {
      conditions.push(`resource_id = $${idx++}`);
      params.push(query.resourceId);
    }
    if (query.tenantId) {
      conditions.push(`tenant_id = $${idx++}`);
      params.push(query.tenantId);
    }
    if (query.dateFrom) {
      conditions.push(`created_at >= $${idx++}::timestamptz`);
      params.push(query.dateFrom);
    }
    if (query.dateTo) {
      conditions.push(`created_at <= $${idx++}::timestamptz`);
      params.push(query.dateTo);
    }

    const where = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

    const countResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM audit_events ${where}`,
      params,
    );
    const total = Number(countResult.rows[0]!.count);

    const dataResult = await this.db.query<AuditRow>(
      `SELECT * FROM audit_events ${where} ORDER BY created_at DESC LIMIT $${idx} OFFSET $${idx + 1}`,
      [...params, pageSize, offset],
    );

    const totalPages = Math.ceil(total / pageSize);
    return {
      data: dataResult.rows,
      meta: { page, pageSize, total, totalPages, hasNext: page < totalPages, hasPrevious: page > 1 },
    };
  }

  async findOne(id: string): Promise<AuditRow> {
    const result = await this.db.query<AuditRow>(`SELECT * FROM audit_events WHERE id = $1`, [id]);
    if (result.rows.length === 0) {
      throw new NotFoundException(`Audit event "${id}" not found`);
    }
    return result.rows[0]!;
  }

  async exportEvents(dto: ExportAuditDto): Promise<{ format: string; content: string; filename: string }> {
    const conditions = ['created_at >= $1::timestamptz', 'created_at <= $2::timestamptz'];
    const params: unknown[] = [dto.dateFrom, dto.dateTo];
    let idx = 3;

    if (dto.tenantId) {
      conditions.push(`tenant_id = $${idx++}`);
      params.push(dto.tenantId);
    }

    const where = `WHERE ${conditions.join(' AND ')}`;
    const result = await this.db.query<AuditRow>(
      `SELECT * FROM audit_events ${where} ORDER BY created_at ASC`,
      params,
    );

    const format = dto.format ?? 'json';
    const filename = `audit-export-${dto.dateFrom}-to-${dto.dateTo}.${format}`;

    if (format === 'csv') {
      const header = 'id,actor_id,action,resource,resource_id,tenant_id,created_at,metadata\n';
      const escaped = (v: string) => `"${v.replace(/"/g, '""')}"`;
      const lines = result.rows.map((r) =>
        [
          r.id,
          r.actor_id,
          r.action,
          r.resource,
          r.resource_id ?? '',
          r.tenant_id ?? '',
          r.created_at.toISOString(),
          escaped(JSON.stringify(r.metadata)),
        ].join(','),
      );
      return { format, content: header + lines.join('\n'), filename };
    }

    return {
      format,
      content: JSON.stringify(result.rows, null, 2),
      filename,
    };
  }

  async complianceReport(query: ComplianceReportQueryDto) {
    const conditions: string[] = [];
    const params: unknown[] = [];
    let idx = 1;

    if (query.tenantId) {
      conditions.push(`tenant_id = $${idx++}`);
      params.push(query.tenantId);
    }
    if (query.dateFrom) {
      conditions.push(`created_at >= $${idx++}::timestamptz`);
      params.push(query.dateFrom);
    }
    if (query.dateTo) {
      conditions.push(`created_at <= $${idx++}::timestamptz`);
      params.push(query.dateTo);
    }

    const where = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

    const byAction = await this.db.query<{ action: string; count: string }>(
      `SELECT action, COUNT(*) AS count FROM audit_events ${where} GROUP BY action ORDER BY count DESC`,
      params,
    );
    const byResource = await this.db.query<{ resource: string; count: string }>(
      `SELECT resource, COUNT(*) AS count FROM audit_events ${where} GROUP BY resource ORDER BY count DESC`,
      params,
    );
    const totalResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) AS count FROM audit_events ${where}`,
      params,
    );

    return {
      byAction: byAction.rows.map((r) => ({ action: r.action, count: Number(r.count) })),
      byResource: byResource.rows.map((r) => ({ resource: r.resource, count: Number(r.count) })),
      total: Number(totalResult.rows[0]!.count),
      generatedAt: new Date(),
    };
  }
}

File: apps/backend/src/backend/services/audit_service.py
import csv
import io
import json
import uuid
from datetime import datetime, timezone

from fastapi import HTTPException
from sqlalchemy import text


class AuditService:
    async def record_event(
        self,
        actor_id: str,
        action: str,
        resource: str,
        resource_id: str | None,
        tenant_id: str | None,
        metadata: dict | None,
        db=None,
    ):
        event_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc)
        await db.execute(
            text("""
                INSERT INTO audit_events (id, actor_id, action, resource, resource_id, tenant_id, metadata, created_at)
                VALUES (:id, :actor_id, :action, :resource, :resource_id, :tenant_id, :metadata, :created_at)
            """),
            {
                "id": event_id,
                "actor_id": actor_id,
                "action": action,
                "resource": resource,
                "resource_id": resource_id,
                "tenant_id": tenant_id,
                "metadata": json.dumps(metadata) if metadata else None,
                "created_at": now,
            },
        )
        return event_id

    async def query_events(
        self,
        page: int,
        page_size: int,
        filters: dict,
        db=None,
    ) -> tuple[list[dict], int]:
        conditions = []
        params: dict = {}

        if filters.get("actor_id"):
            conditions.append("actor_id = :actor_id")
            params["actor_id"] = filters["actor_id"]
        if filters.get("action"):
            conditions.append("action = :action")
            params["action"] = filters["action"]
        if filters.get("resource"):
            conditions.append("resource = :resource")
            params["resource"] = filters["resource"]
        if filters.get("tenant_id"):
            conditions.append("tenant_id = :tenant_id")
            params["tenant_id"] = filters["tenant_id"]
        if filters.get("date_from"):
            conditions.append("created_at >= :date_from")
            params["date_from"] = filters["date_from"]
        if filters.get("date_to"):
            conditions.append("created_at <= :date_to")
            params["date_to"] = filters["date_to"]

        where_clause = " AND ".join(conditions) if conditions else "TRUE"
        offset = (page - 1) * page_size

        count_result = await db.execute(
            text(f"SELECT COUNT(*) FROM audit_events WHERE {where_clause}"),
            params,
        )
        total = count_result.scalar_one() or 0

        rows_result = await db.execute(
            text(f"""
                SELECT id, actor_id, action, resource, resource_id, tenant_id, metadata, created_at
                FROM audit_events
                WHERE {where_clause}
                ORDER BY created_at DESC
                LIMIT :limit OFFSET :offset
            """),
            {**params, "limit": page_size, "offset": offset},
        )
        rows = rows_result.fetchall()

        result = []
        for r in rows:
            meta_raw = r[6]
            if isinstance(meta_raw, str):
                try:
                    meta_raw = json.loads(meta_raw)
                except (json.JSONDecodeError, TypeError):
                    meta_raw = {}
            result.append({
                "id": r[0],
                "actor_id": r[1],
                "action": r[2],
                "resource": r[3],
                "resource_id": r[4],
                "tenant_id": r[5],
                "metadata": meta_raw or {},
                "created_at": r[7].isoformat() if hasattr(r[7], "isoformat") else str(r[7]),
            })

        return result, total

    async def get_event(self, event_id: str, db=None) -> dict | None:
        result = await db.execute(
            text("""
                SELECT id, actor_id, action, resource, resource_id, tenant_id, metadata, created_at
                FROM audit_events WHERE id = :id
            """),
            {"id": event_id},
        )
        r = result.fetchone()
        if not r:
            return None

        meta_raw = r[6]
        if isinstance(meta_raw, str):
            try:
                meta_raw = json.loads(meta_raw)
            except (json.JSONDecodeError, TypeError):
                meta_raw = {}
        return {
            "id": r[0],
            "actor_id": r[1],
            "action": r[2],
            "resource": r[3],
            "resource_id": r[4],
            "tenant_id": r[5],
            "metadata": meta_raw or {},
            "created_at": r[7].isoformat() if hasattr(r[7], "isoformat") else str(r[7]),
        }

    async def export_events(
        self,
        date_from: str | None,
        date_to: str | None,
        format: str,
        tenant_id: str | None,
        db=None,
    ) -> str:
        params: dict = {}
        conditions = []
        if tenant_id:
            conditions.append("tenant_id = :tenant_id")
            params["tenant_id"] = tenant_id
        if date_from:
            conditions.append("created_at >= :date_from")
            params["date_from"] = date_from
        if date_to:
            conditions.append("created_at <= :date_to")
            params["date_to"] = date_to

        where_clause = " AND ".join(conditions) if conditions else "TRUE"

        result = await db.execute(
            text(f"""
                SELECT id, actor_id, action, resource, resource_id, tenant_id, metadata, created_at
                FROM audit_events
                WHERE {where_clause}
                ORDER BY created_at DESC
            """),
            params,
        )
        rows = result.fetchall()

        data = []
        for r in rows:
            meta_raw = r[6]
            if isinstance(meta_raw, str):
                try:
                    meta_raw = json.loads(meta_raw)
                except (json.JSONDecodeError, TypeError):
                    meta_raw = {}
            data.append({
                "id": r[0],
                "actor_id": r[1],
                "action": r[2],
                "resource": r[3],
                "resource_id": r[4],
                "tenant_id": r[5],
                "metadata": meta_raw or {},
                "created_at": r[7].isoformat() if hasattr(r[7], "isoformat") else str(r[7]),
            })

        if format == "csv":
            output = io.StringIO()
            writer = csv.writer(output)
            if data:
                writer.writerow(data[0].keys())
                for row in data:
                    writer.writerow(str(v) for v in row.values())
            return output.getvalue()

        return json.dumps(data, default=str)

    async def compliance_report(
        self,
        tenant_id: str | None,
        date_from: str | None,
        date_to: str | None,
        db=None,
    ) -> dict:
        params: dict = {}
        conditions = []
        if tenant_id:
            conditions.append("tenant_id = :tenant_id")
            params["tenant_id"] = tenant_id
        if date_from:
            conditions.append("created_at >= :date_from")
            params["date_from"] = date_from
        if date_to:
            conditions.append("created_at <= :date_to")
            params["date_to"] = date_to

        where_clause = " AND ".join(conditions) if conditions else "TRUE"

        action_result = await db.execute(
            text(f"""
                SELECT action, COUNT(*) AS cnt
                FROM audit_events
                WHERE {where_clause}
                GROUP BY action
                ORDER BY cnt DESC
            """),
            params,
        )
        action_counts = [{"action": row[0], "count": row[1]} for row in action_result.fetchall()]

        resource_result = await db.execute(
            text(f"""
                SELECT resource, COUNT(*) AS cnt
                FROM audit_events
                WHERE {where_clause}
                GROUP BY resource
                ORDER BY cnt DESC
            """),
            params,
        )
        resource_counts = [{"resource": row[0], "count": row[1]} for row in resource_result.fetchall()]

        total_result = await db.execute(
            text(f"SELECT COUNT(*) FROM audit_events WHERE {where_clause}"),
            params,
        )
        total = total_result.scalar_one() or 0

        return {
            "by_action": action_counts,
            "by_resource": resource_counts,
            "total": total,
            "generated_at": datetime.now(timezone.utc),
        }


audit_service = AuditService()
